src/lightning/pytorch/trainer/supporters.py

```diff
@@ -217,171 +95,91 @@ class CombinedLoader:
-            return
-        length = self._calc_num_batches()
-        state = SharedCycleIteratorState()
-        self.loaders = apply_to_collection(
-            self.loaders, Iterable, CycleIterator, length=length, state=state, wrong_dtype=(Sequence, Mapping)
+        return _map_and_unflatten(
+            lambda x: getattr(x, "batch_sampler", None), self._loaders_flattened, self._loaders_spec
         )
-        state.reset()
-
-    def _apply_cycle_iterator_length(self) -> None:
-        """When the model is `max_size_cycle`, compute the length across all ``CycleIterator`` and re-assign it to
-        all dataloaders."""
-        if self._mode != "max_size_cycle":
-            return
-
-        from lightning.pytorch.utilities.data import get_len
 
-        def set_len(cycle_iterator: CycleIterator, length: int) -> None:
-            cycle_iterator.length = length
-
-        all_lengths = apply_to_collection(self.loaders, CycleIterator, lambda c: get_len(c.loader))
-        max_length = _reduce_data(all_lengths, max, float("-inf"))
-        apply_to_collection(self.loaders, CycleIterator, set_len, length=max_length)
-
-    def __iter__(self) -> Any:
-        """Create and return an iterator, `CombinedLoaderIterator`, for the combined loader."""
-
-        # prevent `NotImplementedError` from PyTorch:
-        # https://github.com/pytorch/pytorch/blob/v1.9.0/torch/utils/data/dataloader.py#L541
-        def __getstate__patch__(*_: Any) -> Dict:
-            return {}
-
-        _BaseDataLoaderIter.__getstate__ = __getstate__patch__  # type: ignore[assignment]
-        iterator = CombinedLoaderIterator(self.loaders)
-        self._iterator = iterator
-        return iterator
-
-    def _calc_num_batches(self) -> _NUMBER:
-        from lightning.pytorch.utilities.data import get_len
-
-        all_lengths = apply_to_collection(self.loaders, Iterable, get_len, wrong_dtype=(Sequence, Mapping))
-        mode = _supported_modes[self._mode]
-        return _reduce_data(all_lengths, mode["fn"], mode["default"])
+    def __next__(self) -> Any:
+        n = len(self._loader_iters)
+        out = [_NoneSentinel()] * n  # values per iterator
+        for i in range(n):
+            try:
+                out[i] = next(self._loader_iters[i])
+            except StopIteration:
+                self._consumed[i] = True
+                if all(self._consumed):
+                    raise StopIteration
+                if self._mode == "max_size_cycle":
+                    # reset the consumed dataloader
+                    self._loader_iters[i] = iter(self._loaders_flattened[i])
+                    out[i] = next(self._loader_iters[i])
+                    continue
+                elif self._mode == "min_size":
+                    raise
+        if any(isinstance(v, _NoneSentinel) for v in out):
+            raise RuntimeError("Something went wrong! Please open an issue on GitHub.")
```

Comment:
I added this for sanity checking, but i'll probably remove it when it's all green